chai_lab/data/parsing/constraints.py

```diff
@@ -0,0 +1,183 @@
+        return int(self.res_idxA[1:])
+
+    @property
+    def res_idxB_name(self) -> str:
+        return self.res_idxB[0]
+
+    @property
+    def res_idxB_pos(self) -> int:
+        return int(self.res_idxB[1:])
+
+    def to_table_entry(self) -> pd.Series:
+        """Format as table entry, sans leading restraint_id column."""
+        values = {
+            "chainA": self.chainA,
+            "res_idxA": (
+                self.res_idxA + (("@" + self.atom_nameA) if self.atom_nameA else "")
+            ),
+            "chainB": self.chainB,
+            "res_idxB": (
+                self.res_idxB + (("@" + self.atom_nameB) if self.atom_nameB else "")
+            ),
+            "connection_type": self.connection_type.value,
+            "confidence": self.confidence,
+            "min_distance_angstrom": self.min_dist_angstrom,
+            "max_distance_angstrom": self.max_dist_angstrom,
+            "comment": self.comment,
+        }
+        return pd.Series(values)
+
+
+def _parse_res_idx(res_idx: str) -> tuple[str, str]:
+    """Parse a residue index string.
+
+    >>> _parse_res_idx("A219")
+    ('A219', '')
+    >>> _parse_res_idx("A219@CA")
+    ('A219', 'CA')
+    >>> _parse_res_idx("@C2")
+    ('', 'C2')
+    """
+    if res_idx.endswith("@"):
+        raise ValueError(f"Invalid residue index: {res_idx}")
+    parts = res_idx.split("@")
+    if len(parts) == 1:
+        parts.append("")
+    elif len(parts) > 2 or all(len(part) == 0 for part in parts):
+        raise ValueError(f"Invalid residue index: {res_idx}")
+    res, atomname = parts
+    return res, atomname
+
+
+def _parse_row(row: pd.Series) -> PairwiseInteraction:
+    """Parse a row of a pairwise interaction table."""
+    resA, atomA = _parse_res_idx(row["res_idxA"])
+    resB, atomB = _parse_res_idx(row["res_idxB"])
+    return PairwiseInteraction(
+        chainA=row["chainA"],
+        res_idxA=resA,
+        atom_nameA=atomA,
+        chainB=row["chainB"],
+        res_idxB=resB,
+        atom_nameB=atomB,
+        max_dist_angstrom=row["max_distance_angstrom"],
+        min_dist_angstrom=row["min_distance_angstrom"],
+        connection_type=PairwiseInteractionType(row["connection_type"]),
+        confidence=row["confidence"],
+        comment=row["comment"],
+    )
+
+
+def parse_pairwise_table(table: str | Path) -> list[PairwiseInteraction]:
+    """Parse a table of pairwise interactions."""
+    df = pd.read_csv(table)
+
+    PairwiseConstraintDataframeModel.validate(df, inplace=True)
+
+    # We only have floats and strings; this applies to strings
+    df.update(df.select_dtypes(include=["object"]).fillna(""))
+
+    # Set confidence to 1.0 if not provided
+    df["confidence"] = df["confidence"].fillna(1.0)
+
+    # Parsing
+    return [_parse_row(row) for _, row in df.iterrows()]
+
+
+def write_pairwise_table(interactions: list[PairwiseInteraction], fname: str | Path):
+    """Write the interactions to a .csv file."""
```

Comment:
parquet? Unless table is expected to be "edited" manually, it better be pqt.

Reply:
This file is expected to be edited manually so I think `.csv` makes more sense here.